components/ws-daemon/pkg/controller/snapshot_controller.go

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/client-go/util/retry"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	workspacev1 "github.com/gitpod-io/gitpod/ws-manager/api/crd/v1"
+)
+
+// SnapshotReconciler reconciles a Snapshot object
+type SnapshotReconciler struct {
+	client.Client
+	nodeName   string
+	operations *WorkspaceOperations
+}
+
+func NewSnapshotController(c client.Client, nodeName string, wso *WorkspaceOperations) *SnapshotReconciler {
+	return &SnapshotReconciler{
+		Client:     c,
+		nodeName:   nodeName,
+		operations: wso,
+	}
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *SnapshotReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&workspacev1.Snapshot{}).
+		WithEventFilter(snapshotEventFilter(r.nodeName)).
```

Comment:
we can check this later, but we'd probably need concurrent reconciles to take snapshots for different workspaces in parallel

Reply:
Yep, will need to check this for backups as well.